### ISSUE TITLE
pilatusimage: header is read twice and incorrectly

### ISSUE THREAD

**Issue by mikapfl** [silx-kit/fabio#198]
Hi,

I noticed that the header returned by tiffIO.getInfo(0) in TifImage.read overwrites the header parsed by PilatusImage._readheader. Therefore, the header parsed in PilatusImage._readheader is never actually used anywhere.

The problem arises because TifImage.read calls self._readheader, which is overridden by PilatusImage, before unconditionally setting the header itself using
```self.header = tiffIO.getInfo(0)```

I don't really know, what the proper fix would be, because I don't know what the API actually should be. In TifImage, it looks like _readheader() should _only_ read the dimensions of the image, not actually the .header property. However, PilatusImage apparently expects that _readheader() reads the whole metadata. One possible fix would be to split the two functionalities into two functions, one _read_dims and one _read_header or such. Or maybe _readheader is actually private to TifImage and PilatusImage should override read() instead?

Cheers

Mika

**Comment by vallsv** [comment]
Thanks for the feed back. Looking for the code, it looks really strange. There is definitely clean up to do here. I don't know why the header is read so many times. Maybe there is an hidden answer there.

But in your case. Is there a problem? Do you expect to read a matadata that is not there? In this case providing a sample and what you expect in the header can help.

**Comment by mikapfl** [comment]
Hi,

well, as the header is read by pilatusimage (but never accessible to calling code), you could get to the important metadata by using:
```python
img.header[b'whatever']
```
(note that keys are binary, not strings, in python3, which should probably also be fixed).
But, with the parsing done with tifimage (actually visible to calling code), I need to use the slightly more awkward
```python
lines = img.header['imageDescription'].decode().split('\r\n')
info = {}
for l in lines:
    key, val = l.split(maxsplit=1)
    info[key[2:]] = val

info['whatever']
```
to get to the same information. Which kinda defeats the purpose of using fabio, since I can have that with approximately the same amount of work directly from PIL with:
```python
from PIL import Image
img = Image(path)
lines = img.tag[270][0].decode().split('\r\n')
info = {}
for l in lines:
    key, val = l.split(maxsplit=1)
    info[key[2:]] = val

info['whatever']
```

Cheers,

Mika

**Comment by mikapfl** [comment]
If you would give me a spec what would be the desirable overall outcome for the pilatusimage reader code, I could maybe prepare a patch, as a first iteration of what needs to be done.

In my case, I can't use fabio at all now, because I need to read the metadata from lzw-compressed .tifs, which TiffIO can't do and since PIL is not used for reading the metadata, I don't get any header information.

I guess one solution could be to override and wrap the TifImage.read() function in PilatusImage.read(), and then we could call TifImage.read() from PilatusImage.read(), let TifImage do all the loading etc., and after that parse the important stuff out of the .header which TifImage already produces. Honestly, the header parsing in PilatusImage looks rather brittle anyway – it will fail as soon as any other tiff header which comes before imageDescription contains '# '.

**Comment by vallsv** [comment]
We don't work with that format, and i dont think we have any spec as usually. That's why i ask for a sample. Maybe we already have something useful around (i found http://www.silx.org/pub/fabio/testimages/lysb_5mg-1.90s_SAXS.bz2).

- You are right, the reader is not robust at all, i will try first to clean up this part. TIFF is AFAIK a well designed format, with typed block of data, it should be easy to read if it was well generated.
- If LZW is needed it can be implemented, there is not problem. But as we don't use this format here, i can't plane anything.
- Multi-frames looks to be non-working, if you have a multi-frame sample it can be useful.

Patches are always welcome, but let me first check if a clean up can be done.

**Comment by vallsv** [comment]
Well, i guess now `_readheader` was designed to be used with `readheader` as a fast way to only read the header.

Unfortunately the result is not constant, when you use `readheader` or `fabio.open`

```
In [10]: import fabio

In [11]: ii = fabio.open("../data/tiff/pilatus/lysb_5mg-1.90s_SAXS")

In [12]: ii
Out[12]: <fabio.pilatusimage.PilatusImage at 0x7faebcad5d90>

In [13]: ii.header
Out[13]: 
{'colormap': None,
 'compression': False,
 'compression_type': 1,
 'date': '2010:11:10 14:05:37\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00',
 'imageDescription': '# Pixel_size 172e-6 m x 172e-6 m\r\n# Silicon sensor, thickness 0.000320 m\r\n# Exposure_time 90.000000 s\r\n# Exposure_period 90.000000 s\r\n# Tau = 0 s\r\n# Count_cutoff 1048574 counts\r\n# Threshold_setting 0 eV\r\n# Gain_setting not implemented (vrf = 9.900)\r\n# N_excluded_pixels = 0\r\n# Excluded_pixels: (nil)\r\n# Flat_field: (nil)\r\n# Trim_directory: (nil)\r\n\x00',
 'info': {},
 'nBits': 32,
 'nColumns': 487,
 'nRows': 619,
 'photometricInterpretation': 1,
 'rowsPerStrip': 619,
 'sampleFormat': 2,
 'software': 'TVX TIFF v 1.3     \x00',
 'stripByteCounts': [1205812],
 'stripOffsets': [4096]}

In [14]: format = fabio.pilatusimage.PilatusImage()

In [15]: format.readheader("../data/tiff/pilatus/lysb_5mg-1.90s_SAXS")

In [16]: format.header
Out[16]: 
{
  "Pixel_size": " 172e-6 m x 172e-6 m", 
  "Silicon": " sensor, thickness 0.000320 m", 
  "Exposure_time": " 90.000000 s", 
  "Exposure_period": " 90.000000 s", 
  "Tau ": " 0 s", 
  "Count_cutoff": " 1048574 counts", 
  "Threshold_setting": " 0 eV", 
  "Gain_setting not implemented (vrf ": " 9.900)", 
  "N_excluded_pixels ": " 0", 
  "Excluded_pixels": " (nil)", 
  "Flat_field": " (nil)", 
  "Trim_directory": " (nil)"
}
```

I think i will only expose the `imageDescription` and store the tiff header somewhere else, just in case.

**Comment by mikapfl** [comment]
Hi,

I'm not using multiframes either. The dectris detectors themselves only write uncompressed single-frame tifs, but I use lzw compression after the measurements to save space.

You can access an lzw-compressed tif from a pilatus here:
[lzw-compressed tif from pilatus](https://oc.basisit.de/index.php/s/BOFgzP6Fn3FRbk0)

Cheers

Mika

**Comment by vallsv** [comment]
Thanks a lot for your sample. @kif looks to be interested about implementing LZW. Is there any problem to use your file under MIT license, we could be able to store it on our test cases.

**Comment by vallsv** [comment]
On the PR #201 i have something working transparently using PIL.
But you have to instantiate the Pilatus image by yourself cause the magic of the file is not detected as a pilatus.

The parsing of the pilatus metadata is not obvious, then let me know if there is something strange.

Hope it is now usable.

```
In [1]: import fabio
In [2]: image = fabio.pilatusimage.PilatusImage()
In [3]: image.read("../data/tiff/pilatus/pilatus_201801_5492.tif")
In [4]: image.header
Out[4]: 
{
  "Pixel_size": "172e-6 m x 172e-6 m", 
  "Silicon": "sensor, thickness 0.000320 m", 
  "Exposure_time": "20.0000000 s", 
  "Exposure_period": "20.0000000 s", 
  "Tau": "800.0e-09 s", 
  "Count_cutoff": "1095546 counts", 
  "Threshold_setting": "3000 eV", 
  "Gain_setting": "ultra high gain (vrf = -0.060)", 
  "N_excluded_pixels": "16", 
  "Excluded_pixels": "badpix_mask.tif", 
  "Flat_field": "FF_p1m0112_E6000_T3000_vrf_m0p06.tif", 
  "Trim_file": "p1m0112_E6000_T3000_vrf_m0p06.bin", 
  "Image_path": "/messung/kmc/daten/2018/1Q/", 
  "Energy_range": "(0, 0) eV", 
  "Detector_distance": "1.00000 m", 
  "Detector_Voffset": "0.00000 m", 
  "Beam_xy": "(0.00, 0.00) pixels", 
  "Flux": "0.000000", 
  "Filter_transmission": "1.0000", 
  "Start_angle": "0.0000 deg.", 
  "Angle_increment": "0.1000 deg.", 
  "Detector_2theta": "0.0000 deg.", 
  "Polarization": "0.990", 
  "Alpha": "0.0000 deg.", 
  "Kappa": "0.0000 deg.", 
  "Phi": "0.0000 deg.", 
  "Chi": "0.0000 deg.", 
  "Oscillation_axis": "X, CW", 
  "N_oscillations": "1"
}

In [5]: image.data
Out[5]: 
array([[2, 1, 2, ..., 6, 5, 2],
       [1, 4, 4, ..., 4, 2, 3],
       [4, 3, 1, ..., 1, 4, 2],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int32)

In [6]: image.data.shape
Out[6]: (1043, 981)

In [7]: image.data.dtype
Out[7]: dtype('int32')
```

**Comment by mikapfl** [comment]
Hi,

you can use the test image however you want. I'd say it is not copyrightable (it's data, not artistic expression, after all), so no license needed.

Cheers,

Mika

**Comment by vallsv** [comment]
Closed by #201
